### PR TITLE
Trim attestation file for deleted resources

### PR DESCRIPTION
**Pull request by Aboli-msft**
## Description
</br>
 Brief description of the requested change.

## Checklist
- [ ] I have read the instructions mentioned in [Contribute to Code](/Contributing.md).
- [ ] I have read and understood the criteria described under [submitting changes](/Contributing.md#submitting-changes). 
- [ ] The title of the PR clearly describes the intent of the PR.
- [ ] This PR does not introduce any breaking changes to the code.

### CODE REVIEW

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 [hunk 1]
```diff
@@ -528,6 +532,135 @@ class ControlStateExtension
...
+				$deletedResourceCount = 0;
+
+				$deletedResourcesWithAttestation = $resourcesWithAttestation| Where-Object{$AzSKScannableResourceIds -notcontains $_ -and $_ -match 'resourceGroups'} # matching with 'resourceGroups' to avoid deleting attestattion for subscription level controls
+				# if any deleted resources found , having attestation for those
+				if($null -ne $deletedResourcesWithAttestation)
+				{		
+					$deletedResourceCount = $deletedResourcesWithAttestation.Count;
+					$filteredIndexerObject = $this.ControlStateIndexer  | Where-Object {$deletedResourcesWithAttestation -notcontains $_.ResourceId}
+					# Rewrite trimmed $this.ControlStateIndexer values to Resource.index.json file
+					[JsonHelper]::ConvertToJsonCustom($filteredIndexerObject) | Out-File $IndexFileLocalTempPath -Force
```
Comment (TarunKrShukla) [inline src/AzSK.Framework/Managers/ControlStateExtension.ps1:634]: Test for single resource

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 [hunk 2]
```diff
@@ -528,6 +532,135 @@ class ControlStateExtension
...
+
+				# Fetch current attestation 
+				$ControlStateIndexerObject = $this.ComputeControlStateIndexer();
+				$resourcesWithAttestation = $this.ControlStateIndexer |Select-Object -ExpandProperty ResourceId
+				
+				$deletedResourcesWithAttestation = @();
+				$filteredIndexerObject= @();
+				$deletedResourceCount = 0;
+
+				$deletedResourcesWithAttestation = $resourcesWithAttestation| Where-Object{$AzSKScannableResourceIds -notcontains $_ -and $_ -match 'resourceGroups'} # matching with 'resourceGroups' to avoid deleting attestattion for subscription level controls
```
Comment (TarunKrShukla) [inline src/AzSK.Framework/Managers/ControlStateExtension.ps1:627]: Check for single object or better put this into an array.

Comment (Aboli-msft) [reply src/AzSK.Framework/Managers/ControlStateExtension.ps1]: Updated as per suggestion

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 (hunk 1 above)
Comment (Aboli-msft) [reply src/AzSK.Framework/Managers/ControlStateExtension.ps1:637]: Updated to array as per suggestion

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 [hunk 3]
```diff
@@ -528,6 +532,135 @@ class ControlStateExtension
...
+			
+			$StorageAccount = $this.AzSKStorageAccount;
+			$IndexFileLocalTempPath= Join-Path $([Constants]::AzSKAppFolderPath) "Temp" | Join-Path -ChildPath $this.UniqueRunId | Join-Path -ChildPath "ServerControlState"| Join-Path -ChildPath $this.IndexerBlobName;
+					
+			# Check if trim has been performed within period specified
+			$PeriodToCheckForLastTrim = [DateTime]::Now.AddDays(-$ControlSettings.AttestationTrimIntervalDays.Days)
+			$BackupFileCount = (Get-AzStorageBlob -Container $ContainerName -Context $StorageAccount.Context -Blob $this.BackupIndexerBlobPrefix | Where-Object { $_.LastModified -gt $PeriodToCheckForLastTrim} | Measure-Object).Count
+			if ($BackupFileCount -gt 0 )
+			{
+				$LastBackup = Get-AzStorageBlob -Container $ContainerName -Context $StorageAccount.Context -Blob $this.BackupIndexerBlobPrefix | Sort-Object -Property LastModified -Descending|Select-Object -First 1
```
[the remark this answers is not in the document]
Comment (Aboli-msft) [reply src/AzSK.Framework/Managers/ControlStateExtension.ps1]: made changes

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 [hunk 4]
```diff
@@ -528,6 +532,137 @@ class ControlStateExtension
...
+		
+		# return if feature flag is off
+		##ToDo: Reversal strategy to restore backup
+		if(!$IsTrimAttestationFeatureEnabled)
+		{
+			return $trimAttestationEvents;
+		}
+		try 
+		{
+			$TrimPeriodExceeded = $false;
```
Comment (VishalHaibatpure) [inline src/AzSK.Framework/Managers/ControlStateExtension.ps1:549]: what/where is the use of $TrimPeriodExceeded flag?

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 [hunk 5]
```diff
@@ -528,6 +532,137 @@ class ControlStateExtension
...
+		{
+			$TrimPeriodExceeded = $false;	
+			$ControlSettings = [ConfigurationManager]::LoadServerConfigFile("ControlSettings.json");
+			
+			$ContainerName = [Constants]::AttestationDataContainerName;
+			
+			#return if you don't have the required state attestation configuration during the runtime evaluation
+			$AttestationContainer= Get-AzStorageContainer -Context $this.AzSKStorageAccount.Context -Name $ContainerName -ErrorAction SilentlyContinue
+			
+			if( $null -eq $this.AzSKResourceGroup -or $null -eq $this.AzSKStorageAccount -or $null -eq $AttestationContainer)
```
Comment (VishalHaibatpure) [inline src/AzSK.Framework/Managers/ControlStateExtension.ps1:557]: Is it possible $this.AzSKStorageAccount will be null at this point as at line 555 we are already using it

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 [hunk 6]
```diff
@@ -528,6 +532,137 @@ class ControlStateExtension
...
+			$StorageAccount = $this.AzSKStorageAccount;
+			$IndexFileLocalTempPath= Join-Path $([Constants]::AzSKAppFolderPath) "Temp" | Join-Path -ChildPath $this.UniqueRunId | Join-Path -ChildPath "ServerControlState"| Join-Path -ChildPath $this.IndexerBlobName;
+					
+			# Check if trim has been performed within period specified
+			$PeriodToCheckForLastTrim = [DateTime]::Now.AddDays(-$ControlSettings.AttestationTrimIntervalDays.Days)
+			$BackUpFiles = Get-AzStorageBlob -Container $ContainerName -Context $StorageAccount.Context -Blob $this.BackupIndexerBlobPrefix | Where-Object { $_.LastModified -gt $PeriodToCheckForLastTrim} 
+			if ($null -ne $BackupFiles)
+			{
+				$LastBackup = $BackUpFiles | Sort-Object -Property LastModified -Descending|Select-Object -First 1 
+				if($LastBackup.LastModified -gt $PeriodToCheckForLastTrim)
```
Comment (VishalHaibatpure) [inline src/AzSK.Framework/Managers/ControlStateExtension.ps1:577]: Does this if statement make sense.. As we have already done it in line no. 573

File: src/AzSK.Framework/Managers/ControlStateExtension.ps1 [hunk 7]
```diff
@@ -528,6 +532,137 @@ class ControlStateExtension
...
+				}
+				$event = "" | Select-Object Name, Properties, Metrics
+				$properties = @{"SubscriptionId"= $this.SubscriptionContext.SubscriptionId;"UniquRunIdentifier"=$this.UniqueRunId;"NumberOfResourcesTrimmed"=$deletedResourcesWithAttestationCount;}
+				$event.Name = "Trim attestation flow completed"
+				$event.Properties = $properties
+				$trimAttestationEvents.Add($event) | Out-Null
+
+				# Purge old backup files
+				$PurgeBackupAttestationFilesBefore = [DateTime]::Now.AddDays(-$ControlSettings.PurgeAttestationbackupAfterDays.Days);
+				$OldBackups = Get-AzStorageBlob -Container $ContainerName -Context $StorageAccount.Context -Blob "$($this.BackupIndexerBlobPrefix)" | Where-Object { $_.LastModified -lt $PurgeBackupAttestationFilesBefore}
```
Comment (VishalHaibatpure) [inline src/AzSK.Framework/Managers/ControlStateExtension.ps1:648]: This double blob fetch should have avoided.

**Pull request merged**